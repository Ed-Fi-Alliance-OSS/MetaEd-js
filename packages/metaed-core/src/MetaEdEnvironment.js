// @flow
import { newEntityRepository } from './model/EntityRepository';
import { newPropertyIndex } from './model/property/PropertyRepository';
import type { EntityRepository } from './model/EntityRepository';
import type { PluginEnvironment } from './plugin/PluginEnvironment';
import type { PropertyIndex } from './model/property/PropertyRepository';

export type SemVer = string;

// provided to validators and enhancers
export type MetaEdEnvironment = {
  // the entity repository
  entity: EntityRepository,

  // the property index, provided for convenience (a little hesitant here, as it should be dependent on entityRepository)
  propertyIndex: PropertyIndex,

  // plugin environment by plugin name
  plugin: Map<string, PluginEnvironment>,

  // the current MetaEd Version
  metaEdVersion: SemVer,

  // the data standard version to target
  dataStandardVersion: SemVer,
};

export const newMetaEdEnvironment: () => MetaEdEnvironment = () => ({
  entity: newEntityRepository(),
  propertyIndex: newPropertyIndex(),
  plugin: new Map(),
  metaEdVersion: '0.0.0',
  dataStandardVersion: '0.0.0',
});
