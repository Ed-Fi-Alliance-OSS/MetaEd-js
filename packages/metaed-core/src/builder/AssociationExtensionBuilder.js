// @flow
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newAssociationExtension } from '../model/AssociationExtension';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { isErrorText } from './BuilderUtility';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates AssociationExtension entities.
 */
export class AssociationExtensionBuilder extends TopLevelEntityBuilder {
  enterAssociationExtension(context: MetaEdGrammar.AssociationExtensionContext) {
    this.enteringEntity(newAssociationExtension);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitAssociationExtension(context: MetaEdGrammar.AssociationExtensionContext) {
    this.exitingEntity();
  }

  enterExtendeeName(context: MetaEdGrammar.ExtendeeNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;

    const extendeeName = context.ID().getText();
    this.enteringName(extendeeName);
    this.currentTopLevelEntity.baseEntityName = extendeeName;
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  }
}
