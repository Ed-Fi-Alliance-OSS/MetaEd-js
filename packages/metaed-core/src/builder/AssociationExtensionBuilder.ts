import type { AssociationExtensionContext, ExtendeeNameContext } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newAssociationExtension } from '../model/AssociationExtension';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { sourceMapFrom } from '../model/SourceMap';

/**
 * An ANTLR4 listener that creates AssociationExtension entities.
 */
export class AssociationExtensionBuilder extends TopLevelEntityBuilder {
  enterAssociationExtension = (context: AssociationExtensionContext) => {
    this.enteringEntity(newAssociationExtension);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  };

  exitAssociationExtension = (_context: AssociationExtensionContext) => {
    this.exitingEntity();
  };

  enterExtendeeName = (context: ExtendeeNameContext) => {
    this.enteringExtendeeName(context);
  };
}
