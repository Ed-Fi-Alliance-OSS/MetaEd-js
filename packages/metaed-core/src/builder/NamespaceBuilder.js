// @flow
import { String as sugar } from 'sugar';
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { MetaEdGrammarListener } from '../grammar/gen/MetaEdGrammarListener';
import type { EntityRepository } from '../model/EntityRepository';
import type { MetaEdEnvironment } from '../MetaEdEnvironment';
import type { Namespace } from '../model/Namespace';
import { NoNamespace, newNamespace } from '../model/Namespace';
import { sourceMapFrom } from '../model/SourceMap';
import { isErrorText } from './BuilderUtility';
import type { ValidationFailure } from '../validator/ValidationFailure';

export function namespaceNameFrom(context: MetaEdGrammar.NamespaceNameContext): string {
  if (
    context.exception ||
    context.NAMESPACE_ID() == null ||
    context.NAMESPACE_ID().exception != null ||
    context.NAMESPACE_ID().getText() == null ||
    isErrorText(context.NAMESPACE_ID().getText())
  )
    return '';

  return context.NAMESPACE_ID().getText();
}

function enteringNamespaceName(context: MetaEdGrammar.NamespaceNameContext, namespace: Namespace): Namespace {
  if (namespace === NoNamespace || isErrorText(context.NAMESPACE_ID().getText())) return newNamespace();

  if (
    context.exception ||
    context.NAMESPACE_ID() == null ||
    context.NAMESPACE_ID().exception != null ||
    context.NAMESPACE_ID().getText() == null
  )
    return namespace;

  Object.assign(namespace, { namespaceName: context.NAMESPACE_ID().getText() });

  // This is a good guess -- capitalize the namespace -- given that projectName is not in the language yet,
  // but the AddProjectNameToNamespace task sets the correct value after the builders run
  namespace.projectName = sugar.capitalize(namespace.namespaceName);

  Object.assign(namespace.sourceMap, { namespaceName: sourceMapFrom(context) });
  return namespace;
}

function enteringNamespaceType(context: MetaEdGrammar.NamespaceTypeContext, namespace: Namespace): Namespace {
  if (namespace === NoNamespace) return namespace;
  if (context.exception) return namespace;
  if (context.CORE() != null) {
    Object.assign(namespace, { projectExtension: '', projectName: 'Ed-Fi', isExtension: false });
    return namespace;
  }
  if (context.ID() == null || context.ID().exception != null || isErrorText(context.ID().getText())) return namespace;

  Object.assign(namespace, {
    projectExtension: context.ID().getText(),
    isExtension: true,
  });
  Object.assign(namespace.sourceMap, { projectExtension: sourceMapFrom(context), isExtension: sourceMapFrom(context) });
  return namespace;
}

export class NamespaceBuilder extends MetaEdGrammarListener {
  entityRepository: EntityRepository;
  currentNamespace: Namespace;
  validationFailures: Array<ValidationFailure>;

  constructor(metaEd: MetaEdEnvironment, validationFailures: Array<ValidationFailure>) {
    super();
    this.entityRepository = metaEd.entity;
    this.currentNamespace = NoNamespace;
    this.validationFailures = validationFailures;
  }

  enterNamespace(context: MetaEdGrammar.NamespaceContext) {
    if (context.exception) return;
    if (this.currentNamespace !== NoNamespace) return;
    this.currentNamespace = newNamespace();
    this.currentNamespace.sourceMap.type = sourceMapFrom(context);
  }

  enterNamespaceName(context: MetaEdGrammar.NamespaceNameContext) {
    if (this.currentNamespace === NoNamespace) return;
    this.currentNamespace = enteringNamespaceName(context, this.currentNamespace);
  }

  enterNamespaceType(context: MetaEdGrammar.NamespaceTypeContext) {
    if (this.currentNamespace === NoNamespace) return;
    this.currentNamespace = enteringNamespaceType(context, this.currentNamespace);
    // we don't wait for exitNamespace to finish building Namespaces
    this.entityRepository.namespace.set(this.currentNamespace.namespaceName, this.currentNamespace);
    this.currentNamespace = NoNamespace;
  }
}
