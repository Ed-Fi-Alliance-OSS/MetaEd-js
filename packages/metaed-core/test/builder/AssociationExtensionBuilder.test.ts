import { AssociationExtensionBuilder } from '../../src/builder/AssociationExtensionBuilder';
import { NamespaceBuilder } from '../../src/builder/NamespaceBuilder';
import { MetaEdTextBuilder } from '../../src/grammar/MetaEdTextBuilder';
import { newMetaEdEnvironment } from '../../src/MetaEdEnvironment';
import { getAssociationExtension } from '../TestHelper';
import { MetaEdEnvironment } from '../../src/MetaEdEnvironment';
import { ValidationFailure } from '../../src/validator/ValidationFailure';

describe('when building association extension in extension namespace', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should build one association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(1);
  });

  it('should be found in entity repository', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName)).toBeDefined();
    expect(getAssociationExtension(namespace.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', (): void => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have extendee name', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityName).toBe(entityName);
  });

  it('should have namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.namespaceName).toBe(namespaceName);
  });

  it('should have base entity namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityNamespaceName).toBe(namespaceName);
  });

  it('should have project extension', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.projectExtension).toBe(projectExtension);
  });

  it('should not be deprecated', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).isDeprecated).toBe(false);
  });

  it('should have one property', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).properties).toHaveLength(1);
  });

  it('should have integer property', (): void => {
    const integerProperty = getAssociationExtension(namespace.entity, entityName).properties[0];

    expect(integerProperty.metaEdName).toBe(propertyName);
    expect(integerProperty.type).toBe('integer');
    expect(integerProperty.isRequired).toBe(true);
  });
});

describe('when building deprecated association extension', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';
  const deprecationReason = 'reason';
  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName)
      .withDeprecated(deprecationReason)
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should build one association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should be deprecated', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).isDeprecated).toBe(true);
    expect(getAssociationExtension(namespace.entity, entityName).deprecationReason).toBe(deprecationReason);
  });
});

describe('when building association extension in extension namespace extending core entity', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const namespaceName = 'Namespace';
  const coreNamespaceName = 'EdFi';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(`${coreNamespaceName}.${entityName}`, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should have extendee name', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityName).toBe(entityName);
  });

  it('should have namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.namespaceName).toBe(namespaceName);
  });

  it('should have base entity namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityNamespaceName).toBe(coreNamespaceName);
  });
});

describe('when building duplicate association extensions', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()

      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should build one association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(1);
  });

  it('should be found in entity repository', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName)).toBeDefined();
    expect(getAssociationExtension(namespace.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have two validation failures', (): void => {
    expect(validationFailures).toHaveLength(2);
  });

  it('should have validation failures for each entity', (): void => {
    expect(validationFailures[0].validatorName).toBe('TopLevelEntityBuilder');
    expect(validationFailures[0].category).toBe('error');
    expect(validationFailures[0].message).toMatchInlineSnapshot(
      `"Association Extension named EntityName is a duplicate declaration of that name."`,
    );
    expect(validationFailures[0].sourceMap).toMatchInlineSnapshot(`
            Object {
              "column": 14,
              "line": 7,
              "tokenText": "EntityName",
            }
        `);

    expect(validationFailures[1].validatorName).toBe('TopLevelEntityBuilder');
    expect(validationFailures[1].category).toBe('error');
    expect(validationFailures[1].message).toMatchInlineSnapshot(
      `"Association Extension named EntityName is a duplicate declaration of that name."`,
    );
    expect(validationFailures[1].sourceMap).toMatchInlineSnapshot(`
            Object {
              "column": 14,
              "line": 2,
              "tokenText": "EntityName",
            }
        `);
  });
});

describe('when building association extension with no association extension name', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = '';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should not build association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(0);
  });

  it('should have no viable alternative error', (): void => {
    expect(textBuilder.errorMessages).toMatchInlineSnapshot(`
            Array [
              "no viable alternative at input 'Associationadditions', column: 15, line: 2, token: additions",
              "no viable alternative at input 'Associationadditions', column: 15, line: 2, token: additions",
            ]
        `);
  });
});

describe('when building association extension with lowercase association extension name', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'entityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should not build association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(0);
  });

  it('should have no viable alternative error', (): void => {
    expect(textBuilder.errorMessages).toMatchInlineSnapshot(`
            Array [
              "no viable alternative at input 'Associatione', column: 14, line: 2, token: e",
              "no viable alternative at input 'Associatione', column: 14, line: 2, token: e",
            ]
        `);
  });
});

describe('when building association extension with no property', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should build one association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(1);
  });

  it('should be found in entity repository', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName)).toBeDefined();
    expect(getAssociationExtension(namespace.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', (): void => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have extendee name', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityName).toBe(entityName);
  });

  it('should have namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.namespaceName).toBe(namespaceName);
  });

  it('should have project extension', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.projectExtension).toBe(projectExtension);
  });

  it('should have no property', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).properties).toHaveLength(0);
  });

  it('should have mismatched input error', (): void => {
    expect(textBuilder.errorMessages).toMatchInlineSnapshot(`
      Array [
        "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'deprecated', METAED_ID}, column: 0, line: 3, token: End Namespace",
        "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'deprecated', METAED_ID}, column: 0, line: 3, token: End Namespace",
      ]
    `);
  });
});

describe('when building association extension with invalid trailing text', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const textBuilder: MetaEdTextBuilder = MetaEdTextBuilder.build();
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  const trailingText = '\r\nTrailingText';
  let namespace: any = null;

  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    textBuilder
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withTrailingText(trailingText)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should build one association extension', (): void => {
    expect(namespace.entity.associationExtension.size).toBe(1);
  });

  it('should be found in entity repository', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName)).toBeDefined();
    expect(getAssociationExtension(namespace.entity, entityName).metaEdName).toBe(entityName);
  });

  it('should have no validation failures', (): void => {
    expect(validationFailures).toHaveLength(0);
  });

  it('should have extendee name', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).baseEntityName).toBe(entityName);
  });

  it('should have namespace', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.namespaceName).toBe(namespaceName);
  });

  it('should have project extension', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).namespace.projectExtension).toBe(projectExtension);
  });

  it('should have one property', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).properties).toHaveLength(1);
  });

  it('should have integer property', (): void => {
    const integerProperty = getAssociationExtension(namespace.entity, entityName).properties[0];

    expect(integerProperty.metaEdName).toBe(propertyName);
    expect(integerProperty.type).toBe('integer');
    expect(integerProperty.isRequired).toBe(true);
  });

  it('should have extraneous input error', (): void => {
    expect(textBuilder.errorMessages).toMatchInlineSnapshot(`
            Array [
              "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'role name'}, column: 0, line: 7, token: TrailingText",
              "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'role name'}, column: 0, line: 7, token: TrailingText",
            ]
        `);
  });
});

describe('when building association extension source map', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const validationFailures: ValidationFailure[] = [];
  const namespaceName = 'Namespace';
  const projectExtension = 'ProjectExtension';

  const entityName = 'EntityName';
  const propertyName = 'PropertyName';
  let namespace: any = null;
  beforeAll(() => {
    const builder = new AssociationExtensionBuilder(metaEd, validationFailures);

    MetaEdTextBuilder.build()
      .withBeginNamespace(namespaceName, projectExtension)
      .withStartAssociationExtension(entityName, '1')
      .withIntegerProperty(propertyName, 'doc', true, false)
      .withEndAssociationExtension()
      .withEndNamespace()
      .sendToListener(new NamespaceBuilder(metaEd, validationFailures))
      .sendToListener(builder);
    namespace = metaEd.namespace.get(namespaceName);
  });

  it('should have a metaEdName property', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).sourceMap.metaEdName).toBeDefined();
  });

  it('should have a type property', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).sourceMap.type).toBeDefined();
  });

  it('should have source map data', (): void => {
    expect(getAssociationExtension(namespace.entity, entityName).sourceMap).toMatchInlineSnapshot(`
      Object {
        "allowPrimaryKeyUpdates": Object {
          "column": 0,
          "line": 0,
          "tokenText": "NoSourceMap",
        },
        "baseEntity": Object {
          "column": 0,
          "line": 0,
          "tokenText": "NoSourceMap",
        },
        "baseEntityName": Object {
          "column": 14,
          "line": 2,
          "tokenText": "EntityName",
        },
        "baseEntityNamespaceName": Object {
          "column": 14,
          "line": 2,
          "tokenText": "EntityName",
        },
        "deprecationReason": Object {
          "column": 0,
          "line": 0,
          "tokenText": "NoSourceMap",
        },
        "documentation": Object {
          "column": 0,
          "line": 0,
          "tokenText": "NoSourceMap",
        },
        "identityProperties": Array [],
        "isDeprecated": Object {
          "column": 0,
          "line": 0,
          "tokenText": "NoSourceMap",
        },
        "metaEdName": Object {
          "column": 14,
          "line": 2,
          "tokenText": "EntityName",
        },
        "properties": Array [],
        "queryableFields": Array [],
        "type": Object {
          "column": 2,
          "line": 2,
          "tokenText": "Association",
        },
      }
    `);
  });
});
