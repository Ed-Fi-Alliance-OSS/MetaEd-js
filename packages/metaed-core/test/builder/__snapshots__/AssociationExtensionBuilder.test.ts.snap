// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when building association extension source map should have source map data 1`] = `
Object {
  "allowPrimaryKeyUpdates": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "baseEntity": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "baseEntityName": Object {
    "column": 14,
    "line": 2,
    "tokenText": "EntityName",
  },
  "baseEntityNamespaceName": Object {
    "column": 14,
    "line": 2,
    "tokenText": "EntityName",
  },
  "documentation": Object {
    "column": 0,
    "line": 0,
    "tokenText": "NoSourceMap",
  },
  "identityProperties": Array [],
  "metaEdId": Object {
    "column": 35,
    "line": 2,
    "tokenText": "[1]",
  },
  "metaEdName": Object {
    "column": 14,
    "line": 2,
    "tokenText": "EntityName",
  },
  "properties": Array [],
  "queryableFields": Array [],
  "type": Object {
    "column": 2,
    "line": 2,
    "tokenText": "Association",
  },
}
`;

exports[`when building association extension with invalid trailing text should have extraneous input error 1`] = `
Array [
  "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'with context'}, column: 0, line: 7, token: TrailingText",
  "extraneous input 'TrailingText' expecting {'Abstract Entity', 'Association', 'End Namespace', 'Choice', 'Common', 'Descriptor', 'Domain', 'Domain Entity', 'Enumeration', 'Interchange', 'Inline Common', 'Shared Decimal', 'Shared Integer', 'Shared Short', 'Shared String', 'Subdomain', 'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year', 'is queryable field', 'min value', 'max value', 'with context'}, column: 0, line: 7, token: TrailingText",
]
`;

exports[`when building association extension with lowercase association extension name should have no viable alternative error 1`] = `
Array [
  "no viable alternative at input 'Associatione', column: 14, line: 2, token: e",
  "no viable alternative at input 'Associatione', column: 14, line: 2, token: e",
]
`;

exports[`when building association extension with no association extension name should have no viable alternative error 1`] = `
Array [
  "no viable alternative at input 'Associationadditions', column: 15, line: 2, token: additions",
  "no viable alternative at input 'Associationadditions', column: 15, line: 2, token: additions",
]
`;

exports[`when building association extension with no property should have mismatched input error 1`] = `
Array [
  "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year'}, column: 0, line: 3, token: End Namespace",
  "mismatched input 'End Namespace' expecting {'association', 'bool', 'choice', 'common', 'common extension', 'currency', 'date', 'datetime', 'decimal', 'descriptor', 'domain entity', 'duration', 'enumeration', 'inline common', 'integer', 'percent', 'shared decimal', 'shared integer', 'shared short', 'shared string', 'short', 'string', 'time', 'year'}, column: 0, line: 3, token: End Namespace",
]
`;

exports[`when building duplicate association extensions should have validation failures for each entity: when building duplicate association extensions should have validation failures for each entity -> Association 1 message 1`] = `"Association Extension named EntityName is a duplicate declaration of that name."`;

exports[`when building duplicate association extensions should have validation failures for each entity: when building duplicate association extensions should have validation failures for each entity -> Association 1 sourceMap 1`] = `
Object {
  "column": 14,
  "line": 7,
  "tokenText": "EntityName",
}
`;

exports[`when building duplicate association extensions should have validation failures for each entity: when building duplicate association extensions should have validation failures for each entity -> Association 2 message 1`] = `"Association Extension named EntityName is a duplicate declaration of that name."`;

exports[`when building duplicate association extensions should have validation failures for each entity: when building duplicate association extensions should have validation failures for each entity -> Association 2 sourceMap 1`] = `
Object {
  "column": 14,
  "line": 2,
  "tokenText": "EntityName",
}
`;
